pkg/router/controller/unique_host.go

```diff
@@ -179,6 +181,9 @@ func (p *UniqueHost) HandleRoute(eventType watch.EventType, route *routev1.Route
 			// we were not added because another route is covering us
 			var owner *routev1.Route
 			if old, ok := p.index.RoutesForHost(host); ok && len(old) > 0 {
+				sort.SliceStable(old, func(i, j int) bool {
+					return !routeapihelpers.RouteLessThan(old[i], old[j])
+				})
```

Comment:
I'm not entirely sure this fixes the issue. `old` here seems to be all the routes with the same host, but I think we also need to take path into account to decide which route conflicts. Sorting all the routes works for [the reproducer script from your haproxy-hacks repo](https://github.com/frobware/haproxy-hacks/blob/master/OCPBUGS-16707/reproducer/repro.sh), but if I edit the script to create the routes in the order route2 > route1 > route3, then route3 reports that route1 is the conflicting route, even though it's actually route2 that's the problem:
```
route.route.openshift.io/route2 created
route.route.openshift.io/route1 created
route.route.openshift.io/route3 created
Name:                   route3
Namespace:              ocpbugs16707
Created:                3 seconds ago
Labels:                 <none>
Annotations:            <none>
Requested Host:         httpd-example-path-based-routes.apps.firstcluster.lab.upshift.rdu2.redhat.com
                        rejected by router default:  (host router-default.apps.ci-ln-xxi6972-72292.gcp-2.ci.openshift.org)HostAlreadyClaimed (3 seconds ago)
                          route route1 already exposes httpd-example-path-based-routes.apps.firstcluster.lab.upshift.rdu2.redhat.com and is older
Path:                   /api
TLS Termination:        <none>
Insecure Policy:        <none>
Endpoint Port:          http

Service:        api-service
Weight:         100 (100%)
Endpoints:      <error: endpoints "api-service" not found>
```

Reply:
I pushed an update in https://github.com/openshift/router/commit/ca357ed0bd537bc7f01ce71e46a6b50f6f2b18a8.